packages/devtools_app/lib/src/shared/preferences.dart

```diff
@@ -114,25 +129,71 @@ class InspectorPreferencesController extends DisposableController
     });
   }
 
-  void _initCustomPubRootListeners() {
-    // TODO(CoderDake): add _customPubRootDirectories listener back when
-    // finalizing https://github.com/flutter/devtools/issues/3941
-    /*
-    addAutoDisposeListener(_customPubRootDirectories, () {
-      storage.setValue(
-        _customPubRootDirectoriesStorageId,
-        jsonEncode(_customPubRootDirectories.value),
-      );
-    });
-    */
+  Future<void> _initCustomPubRootDirectories() async {
+    autoDisposeStreamSubscription(
+      serviceManager.onConnectionAvailable
+          .listen(_handleConnectionToNewService),
+    );
+    autoDisposeStreamSubscription(
+      serviceManager.onConnectionClosed.listen(_handleConnectionClosed),
+    );
     addAutoDisposeListener(_busyCounter, () {
       _customPubRootDirectoriesAreBusy.value = _busyCounter.value != 0;
     });
+    addAutoDisposeListener(
+      serviceManager.isolateManager.mainIsolate,
+      () {
+        if (_mainScriptDir != null &&
+            serviceManager.isolateManager.mainIsolate.value != null) {
+          final debuggerState =
+              serviceManager.isolateManager.mainIsolateDebuggerState;
+
+          if (debuggerState?.isPaused.value == false) {
+            // the isolate is already unpaused, we can try to load
+            // the directories
+            preferences.inspector.loadCustomPubRootDirectories();
+          } else {
+            late Function() pausedListener;
+
+            pausedListener = () {
+              if (debuggerState?.isPaused.value == false) {
+                preferences.inspector.loadCustomPubRootDirectories();
+
+                debuggerState?.isPaused.removeListener(pausedListener);
```

Comment:
we should override `dispose` on `InspectorPreferencesController` and remove this listener. That way, in case the app is never unpaused this listener will still be removed.

Reply:
I think just adding autodispose for the listener will accomplish this.
It has the added benefit of still removing listeners that are around, even if the main isolate switches around a bunch too!